File: Libs/pollilib/javascript/polliLib/vision.js
import fs from 'node:fs';

export const VisionMixin = (Base) => class extends Base {
  async analyze_image_url(imageUrl, options = {}) {
    const {
      question = "What's in this image?",
      model = 'openai',
      max_tokens = 500,
      referrer = null,
      token = null,
      timeoutMs,
      asJson = false,
    } = options;
    const payload = {
      model,
      messages: [ { role: 'user', content: [ { type: 'text', text: question }, { type: 'image_url', image_url: { url: imageUrl } } ] } ],
    };
    if (typeof max_tokens === 'number') payload.max_tokens = max_tokens;
    if (referrer) payload.referrer = referrer;
    if (token) payload.token = token;
    payload.safe = false;
    const url = `${this.textPromptBase}/${model}`;
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), this._resolveTimeout(timeoutMs, 60_000));
    try {
      const resp = await this.fetch(url, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(payload), signal: controller.signal });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const json = await resp.json();
      if (asJson) return json;
      return json?.choices?.[0]?.message?.content;
    } finally { clearTimeout(t); }
  }

  async analyze_image_file(imagePath, options = {}) {
    const {
      question = "What's in this image?",
      model = 'openai',
      max_tokens = 500,
      referrer = null,
      token = null,
      timeoutMs,
      asJson = false,
    } = options;
    if (!fs.existsSync(imagePath)) throw new Error(`File not found: ${imagePath}`);
    let ext = String(imagePath).split('.').pop().toLowerCase();
    if (!['jpeg','jpg','png','gif','webp'].includes(ext)) ext = 'jpeg';
    const data = await fs.promises.readFile(imagePath);
    const b64 = data.toString('base64');
    const dataUrl = `data:image/${ext};base64,${b64}`;
    const payload = {
      model,
      messages: [ { role: 'user', content: [ { type: 'text', text: question }, { type: 'image_url', image_url: { url: dataUrl } } ] } ],
    };
    if (typeof max_tokens === 'number') payload.max_tokens = max_tokens;
    if (referrer) payload.referrer = referrer;
    if (token) payload.token = token;
    payload.safe = false;
    const url = `${this.textPromptBase}/${model}`;
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), this._resolveTimeout(timeoutMs, 60_000));
    try {
      const resp = await this.fetch(url, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(payload), signal: controller.signal });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const json = await resp.json();
      if (asJson) return json;
      return json?.choices?.[0]?.message?.content;
    } finally { clearTimeout(t); }
  }
};

File: Libs/pollilib/python/polliLib/vision.py
from __future__ import annotations

from typing import Any, Dict, Optional


class VisionMixin:
    def analyze_image_url(
        self,
        image_url: str,
        *,
        question: str = "What's in this image?",
        model: str = "openai",
        max_tokens: Optional[int] = 500,
        referrer: Optional[str] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = None,
        as_json: bool = False,
    ) -> Any:
        payload: Dict[str, Any] = {
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {"type": "text", "text": question},
                        {"type": "image_url", "image_url": {"url": image_url}},
                    ],
                }
            ],
        }
        if max_tokens is not None:
            payload["max_tokens"] = int(max_tokens)
        if referrer:
            payload["referrer"] = referrer
        if token:
            payload["token"] = token
        payload["safe"] = False
        url = f"{self.text_prompt_base}/{model}"
        headers = {"Content-Type": "application/json"}
        eff_timeout = self._resolve_timeout(timeout, 60.0)
        resp = self.session.post(url, headers=headers, json=payload, timeout=eff_timeout)
        resp.raise_for_status()
        data = resp.json()
        if as_json:
            return data
        return data.get("choices", [{}])[0].get("message", {}).get("content")

    def analyze_image_file(
        self,
        image_path: str,
        *,
        question: str = "What's in this image?",
        model: str = "openai",
        max_tokens: Optional[int] = 500,
        referrer: Optional[str] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = None,
        as_json: bool = False,
    ) -> Any:
        import os, base64
        if not os.path.exists(image_path):
            raise FileNotFoundError(image_path)
        ext = os.path.splitext(image_path)[1].lower().lstrip(".")
        if ext not in {"jpeg", "jpg", "png", "gif", "webp"}:
            ext = "jpeg"
        with open(image_path, "rb") as f:
            b64 = base64.b64encode(f.read()).decode("utf-8")
        data_url = f"data:image/{ext};base64,{b64}"
        payload: Dict[str, Any] = {
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {"type": "text", "text": question},
                        {"type": "image_url", "image_url": {"url": data_url}},
                    ],
                }
            ],
        }
        if max_tokens is not None:
            payload["max_tokens"] = int(max_tokens)
        if referrer:
            payload["referrer"] = referrer
        if token:
            payload["token"] = token
        payload["safe"] = False
        url = f"{self.text_prompt_base}/{model}"
        headers = {"Content-Type": "application/json"}
        eff_timeout = self._resolve_timeout(timeout, 60.0)
        resp = self.session.post(url, headers=headers, json=payload, timeout=eff_timeout)
        resp.raise_for_status()
        data = resp.json()
        if as_json:
            return data
        return data.get("choices", [{}])[0].get("message", {}).get("content")

